server/auth/oauth2/service.go

```diff
@@ -0,0 +1,147 @@
+	}
+	idTokenVerifier := provider.Verifier(&oidc.Config{ClientID: clientID})
+	log.WithFields(log.Fields{"redirectURL": config.RedirectURL, "issuer": issuer, "clientId": clientID}).Info("SSO configuration")
+	return &service{config, idTokenVerifier, baseHRef}, nil
+}
+
+const stateCookieName = "oauthState"
+
+func (s *service) HandleRedirect(w http.ResponseWriter, r *http.Request) {
+	state := rand.RandString(10)
+	// TODO - no path?
+	http.SetCookie(w, &http.Cookie{
+		Name:     stateCookieName,
+		Value:    state,
+		Expires:  time.Now().Add(1 * time.Hour),
+		HttpOnly: true,
+		// TODO - lax? not strict?
+		SameSite: http.SameSiteLaxMode,
+	})
+	http.Redirect(w, r, s.config.AuthCodeURL(state), http.StatusFound)
+}
+
+func (s *service) HandleCallback(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	state := r.URL.Query().Get("state")
+	cookie, err := r.Cookie(stateCookieName)
+	http.SetCookie(w, &http.Cookie{Name: stateCookieName, MaxAge: 0})
+	if err != nil {
+		w.WriteHeader(400)
+		_, _ = w.Write([]byte(fmt.Sprintf("invalid state: %v", err)))
+		return
+	}
+	if state != cookie.Value {
+		w.WriteHeader(401)
+		_, _ = w.Write([]byte(fmt.Sprintf("invalid state: %s", state)))
+		return
+	}
+	oauth2Token, err := s.config.Exchange(ctx, r.URL.Query().Get("code"))
+	if err != nil {
+		w.WriteHeader(401)
+		_, _ = w.Write([]byte(fmt.Sprintf("failed to exchange token: %v", err)))
+		return
+	}
+	rawIDToken, ok := oauth2Token.Extra("id_token").(string)
+	if !ok {
+		w.WriteHeader(401)
+		_, _ = w.Write([]byte(fmt.Sprintf("failed to get id_token")))
+		return
+	}
+	idToken, err := s.idTokenVerifier.Verify(ctx, rawIDToken)
+	if err != nil {
+		w.WriteHeader(401)
+		_, _ = w.Write([]byte(fmt.Sprintf("failed to verify token: %v", err)))
+		return
+	}
+	c := &claims{}
+	if err := idToken.Claims(c); err != nil {
+		w.WriteHeader(401)
+		_, _ = w.Write([]byte(fmt.Sprintf("failed to get claims: %v", err)))
+		return
+	}
+	value := prefix + rawIDToken
+	log.Debugf("handing oauth2 callback %v", value)
+	// TODO MaxAge? Expires? HttpOnly?
+	// TODO we must compress this because we know id_token can be large if you have many groups
+	http.SetCookie(w, &http.Cookie{Name: "authorization", Value: value, Path: s.baseHRef, SameSite: http.SameSiteStrictMode})
+	http.Redirect(w, r, s.baseHRef, 302)
+}
+
+// authorize verifies a bearer token and pulls user information form the claims.
+func (s *service) Authorize(ctx context.Context, authorisation string) (wfv1.User, error) {
+	rawIDToken := strings.TrimPrefix(authorisation, prefix)
+	idToken, err := s.idTokenVerifier.Verify(ctx, rawIDToken)
+	if err != nil {
+		return wfv1.NullUser, fmt.Errorf("failed to verify id_token %v", err)
+	}
+	c := &claims{}
+	if err := idToken.Claims(c); err != nil {
+		return wfv1.NullUser, fmt.Errorf("failed to parse claims: %v", err)
+	}
+	return wfv1.User{Name: idToken.Subject, Groups: c.Groups}, nil
+}
+
+func (s *service) IsSSO(authorization string) bool {
```

Comment:
I don't agree that this method should exist. A JWT is a JWT. It reasonable for a JWT to be the complete contents of the auth header. If you need to determine whether or not it is SSO (versus kubernetes token), this can be determined by the issuer of the JWT. Not through some invented auth token prefix..

Reply:
Are you saying if (a) the token does not start with `Bearer` or `Basic` and (b) it is a JWET then assume it is SSO?

Reply:
No, the k8s and the SSO tokens are both `Bearer` tokens, both in the form of a JWT. The problem, is how to distinguish between a JWT issued by K8s versus SSO. As I understand your implementation, you look at a specialized prefix. However, I am suggesting you can just look at the issuer, rather than rely on a specialized prefix.

I should rephrase that `IsSSO()` can exist, but it would be changed to unmarshal the JWT to switch on the issuer.

Reply:
Ah. No, I think you maybe wrong on this. To clarify, we accept an "authorization" header. Today,  can be one of the following:

1. Empty string.
2. "Bearer ${bearerToken}" - i.e. K8S 
3. "Basic base64(${username}:${password)" - i.e. K8S

We need a way to determine what kind of token it is - but it is our token - it can whatever we want. I think bearer tokens are  opaque - we don't know if it is JWT or other - this is up to the Kubernetes provider.

Either way, it does not matter. The "authorization" header can be whatever we want, but we should be able to inspect it and make determination - we don't need to be smart about it - a dumb prefix will be simple and robust.

Hence this func:

https://github.com/argoproj/argo/pull/2745/files#diff-8ac7522dba311812e435bdd519739dddR35

Reply:
As promised I revisited this and found a third-way. Authorization header can now be the following:

* `Bearer id_token:${idToken}`
* `Bearer ${k8sBearerToken}`
* `Basic ${k8sBasic}`

This is both simple (no complex and costly code to unpack token) and standardized (what @jessesuen wanted)